jobrunner/run.py

```diff
@@ -66,19 +66,40 @@ def handle_jobs(api: Optional[ExecutorAPI]):
     log.debug("Querying database for active jobs")
     active_jobs = find_where(Job, state__in=[State.PENDING, State.RUNNING])
     log.debug("Done query")
-    # Randomising the job order is a crude but effective way to ensure that a
-    # single large job request doesn't hog all the workers. We make this
-    # optional as, when running locally, having jobs run in a predictable order
-    # is preferable
-    if config.RANDOMISE_JOB_ORDER:
-        random.shuffle(active_jobs)
-
-    for job in active_jobs:
+
+    running_for_workspace = collections.defaultdict(int)
+    handled_jobs = []
+
+    while active_jobs:
+        # We need to re-sort on each loop because the number of running jobs per
+        # workspace will change as we work our way through
+        active_jobs.sort(
+            key=lambda job: (
+                # Process all running jobs first. Once we've processed all of these, the
+                # counts in `running_for_workspace` will be up-to-date.
+                0 if job.state == State.RUNNING else 1,
+                # Then process PENDING jobs in order of how many are running in the
+                # workspace. This gives a fairer allocation of capacity among
+                # workspaces.
+                running_for_workspace[job.workspace],
+                # Finally use job age as a tie-breaker
+                job.created_at,
+            )
+        )
+        job = active_jobs.pop(0)
+
         # `set_log_context` ensures that all log messages triggered anywhere
         # further down the stack will have `job` set on them
         with set_log_context(job=job):
             handle_single_job(job, api)
-    return active_jobs
+
+        # Add running jobs to the workspace count
+        if job.state == State.RUNNING:
+            running_for_workspace[job.workspace] += 1
```

Comment:
should this be `if job.state == State.RUNNING and job_previous_state != State.RUNNING`?

Reply:
Good question, but I think the answer's no – and this is one of the simplifications over the previous version. We don't count running jobs at all at the start, we just build up the count as we go along. That means the count isn't accurate during the initial phase while we processing running jobs, but that's fine. By the time we get to processing the pending jobs it will be accurate.

And because we only process each job once we never risk double-counting. So it makes the logic here much simpler.

Reply:
yep yep, I still had the previous version of this in mind, this is correct.